src/V3Timing.cpp

```diff
@@ -0,0 +1,785 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Prepare AST for timing features
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2022 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+// There are two main steps in V3Timing:
+// - timingAll() which invokes TimingVisitor
+// - toCStmts() which invokes TimingToCStmtsVisitor
+//
+// TimingVisitor transformations:
+// - for each delay, scale it according to the module's timescale
+// - for each inter assignment timing control:
+//     - introduce an intermediate variable
+//     - write the original RHS to the intermediate variable before the timing control
+//     - write the intermediate variable to the original LHS after the timing control
+// - for each process or C++ function, if it has awaiting statements (such as event controls,
+// delays, waits), mark it as suspendable
+//     - if we mark a virtual function as suspendable, mark all overriding and overridden functions
+//     as suspendable
+//     - if we mark an 'always' process as suspendable:
+//         - if there is a sentree, prevent the process from being triggered multiple times
+//         concurrently:
+//             - add a new bool variable that has to be true for the process to run
+//             - if the process is run, set it to false
+//             - set it to true at the end
+//         - if there is no sentree, replace it with an 'initial forever' process
+//
+// TimingToCStmtsVisitor transformations:
+// - create a global delay scheduler (see verilated_timing.h)
+// - replace each delay with an await statement waiting on the global delay scheduler (with the
+// specified delay)
+// - for each event control:
+//     - if there's no corresponding trigger scheduler (see verilated_timing.h), create it
+//     - replace with an await statement waiting on the corresponding trigger scheduler
+// - for each fork, transform each statement to a separate suspendable function
+// - create additional necessary statements to inject into the global eval
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3EmitV.h"
+#include "V3Graph.h"
+#include "V3SenTree.h"
+#include "V3Timing.h"
+#include "V3UniqueNames.h"
+
+//######################################################################
+// Transform nodes affected by timing
+
+class TimingVisitor final : public VNVisitor {
+private:
+    // TYPES
+    using VarScopeSet = std::unordered_set<AstVarScope*>;
```

Comment:
You use this in one place, so inline it please. Also unordered set has non-deterministic iteration order, so use a vector instead.

Reply:
No longer relevant